Implement JITless Wasm test mode

danlliu opened the pull request:
#### 43d87cd9cf468761d5f4cde71c81b6e0808324f7
<pre>
Implement JITless Wasm test mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=281698">https://bugs.webkit.org/show_bug.cgi?id=281698</a>
<a href="https://rdar.apple.com/137215038">rdar://137215038</a>

Reviewed by Keith Miller.

This patch provides a `--useWasmJIT` test mode, which specifically disables JIT for Wasm
functions. This allows us to test JITless Wasm more effectively, since the JS wrappers
will not be slowed down. This can be optionally enabled in `run-javascriptcore-tests`
using the `--jitless-wasm` flag.

* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::disableAllWasmJITOptions):
(JSC::disableAllJITOptions):
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::JSEntrypointCallee::JSEntrypointCallee):
* Source/JavaScriptCore/wasm/WasmEntryPlan.cpp:
(JSC::Wasm::EntryPlan::generateWasmToWasmStubs):
(JSC::Wasm::EntryPlan::generateWasmToJSStubs):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.cpp:
(JSC::Wasm::IPIntPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::compileFunction):

Canonical link: <a href="https://commits.webkit.org/285672@main">https://commits.webkit.org/285672@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56b051abcd1e9a2c5fc1b054c6f29ef63797a794

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73427 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52856 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26235 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77684 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24665 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75542 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/61985 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/661 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57702 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16146 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76494 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47782 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63176 "Found 1 new API test failure: TestWebKitAPI.HSTS.Preconnect (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38113 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44393 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20658 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22996 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66562 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66204 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21011 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79310 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72684 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/743 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/259 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66106 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/884 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63188 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65384 "Found 1 new API test failure: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/open-window-default-size (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16168 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9223 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7407 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94464 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/706 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20764 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/736 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/720 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/738 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/Options.cpp:
```diff
@@ -650,10 +650,25 @@ static inline void disableAllWasmOptions()
 
+static inline void disableAllWasmJITOptions()
+{
+    Options::useLLInt() = true;
```
Can we have `disableAllWasmOptions()` call this instead of duplicating the logic?

kmiller68 commented on Source/JavaScriptCore/runtime/Options.cpp:
```diff
@@ -775,6 +791,9 @@ void Options::notifyOptionsChanged()
     if (!Options::useWasm() || !Options::useJIT())
         disableAllWasmOptions();
 
+    if (!Options::useWasmJIT())
```
I think we should do 
```
if (!Options::useJIT()) 
    Options::useWasmJIT() = false;
if (!Options::useWasmJIT())
    disableAllWasmJITOptions();
``` 
then we only have to do `Options::useWasmJIT()` in the runtime.

hyjorc1 reviewed (dismissed):
r=me

kmiller68 commented on JSTests/wasm/gc/simd.js:
```diff
@@ -1,4 +1,5 @@
 //@ skip unless $isSIMDPlatform
+//@ requireOptions("--useWasmSIMD=1")
```
Is this right? Shouldn't these force `--useWasmJIT=1` until we have IPInt working?

MenloDorian replied:
I think Daniel is using `--useWasmSIMD=1` as an annotation (as a hack) to exclude these tests (with some pre-processing).  I think it would be wrong to require `--useWasmJIT=1` because it would communicate that this test must be run with the JIT, which eventually, it won't be.

However, it we don't need `--useWasmJIT=1` for anything else, then it can be the hack annotation, and we can don't need a preprocessor to filter these out.

kmiller68 replied:
I see... I'm not sure that's any less confusing or more likely to ensure we remove the "skip". IMO, we should just add a `def annotateSIMDTest` to `run-jsc-stress-tests` that disables the test in `no-wasm-jit` config. It should probably have a comment to remove `annotateSIMDTest` once we enable IPInt SIMD, as `annotateSIMDTest` stands out and is easy to grep for.

MenloDorian replied:
The advantage of `--useWasmSIMD=1` is that many relevant tests are already annotated so.  And it does correct ly document the test's expectation.  I think this is an acceptable hack for the short term.

kmiller68 replied:
Yeah, I don't feel too strongly. Although, I think we should file a radar to remove the hack and block it on IPInt.

kmiller68 reviewed (changes requested):
I think we need `Options::useWasmJIT()` checks in JSToWasm/WasmToJS too?

kmiller68 commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -454,7 +454,7 @@ JSEntrypointCallee::JSEntrypointCallee(TypeIndex typeIndex, bool usesSIMD)
 
 #if ENABLE(JIT)
-    if (Options::useJIT()) {
+    if (Options::useJIT() && Options::useWasmJIT()) {
```
These could all be just `if (Options::useWasmJIT())` I think.

kmiller68 reviewed (approved):
r=me with `if (Options::useJIT() && Options::useWasmJIT())` => `if (Options::useWasmJIT())` change.

danlliu committed:
Implement JITless Wasm test mode
https://bugs.webkit.org/show_bug.cgi?id=281698
rdar://137215038

Reviewed by Keith Miller.

This patch provides a `--useWasmJIT` test mode, which specifically disables JIT for Wasm
functions. This allows us to test JITless Wasm more effectively, since the JS wrappers
will not be slowed down. This can be optionally enabled in `run-javascriptcore-tests`
using the `--jitless-wasm` flag.

* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::disableAllWasmJITOptions):
(JSC::disableAllJITOptions):
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::JSEntrypointCallee::JSEntrypointCallee):
* Source/JavaScriptCore/wasm/WasmEntryPlan.cpp:
(JSC::Wasm::EntryPlan::generateWasmToWasmStubs):
(JSC::Wasm::EntryPlan::generateWasmToJSStubs):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.cpp:
(JSC::Wasm::IPIntPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::compileFunction):

Canonical link: https://commits.webkit.org/285672@main
